arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/stream/ops/SwitchMapTest.kt

```diff
@@ -0,0 +1,163 @@
+  "flatMap equivalence when switching never occurs" {
+    checkAll(Arb.stream(Arb.int())) { s ->
+      val expected = s.compile().toList()
+      Stream
+        .effect { Semaphore(1) }
+        .flatMap { guard ->
+          // wait for inner to emit to prevent switching
+          s.effectTap { guard.acquire() }
+            // outer terminates, wait for last inner to emit
+            .onFinalize { guard.acquire() }
+            .switchMap { x -> Stream.emits(x).onFinalize { guard.release() } }
+        }
+        .compile()
+        .toList() shouldBe expected
+    }
+  }
+
+  "inner stream finalizer always runs before switching" {
+    checkAll(Arb.stream(Arb.int())) { s ->
+      Stream.effect {
+        val ref = Atomic(true)
+        s.switchMap {
+          Stream.effect {
+            ref.get()
+          }.flatMap { released ->
+            if (!released) Stream.raiseError(Throwable())
+            else
+              Stream.effect {
+                ref.set(false)
+                sleep(20.milliseconds)
+              }
+                .onFinalize {
+                  sleep(100.milliseconds)
+                  ref.set(true)
+                }
+          }
+        }
+      }
+        .compile()
+        .drain()
+    }
+  }
+
+  "when primary stream terminates, inner stream continues"  {
+    checkAll(Arb.stream(Arb.int()), Arb.stream(Arb.int())) { s1, s2 ->
+      val expected = s1.last().terminateOnNull().flatMap { s -> s2.append { Stream(s) } }.compile().toList()
+
+      s1.switchMap { s -> Stream.sleep_(20.milliseconds).append { s2 }.append { Stream(s) } }
+        .compile()
+        .toList() shouldBe expected
+    }
+  }
+
+  "when inner stream fails, overall stream fails"  {
+    checkAll(Arb.stream(Arb.int()), Arb.throwable()) { s0, e ->
+      val s = Stream(0).append { s0 }
+        .delayBy(25.milliseconds)
+        .flatMap { Stream.raiseError<Int>(e) }
+        .compile()
+
+      assertThrowable {
+        s.drain()
+      } shouldBe e
+    }
+  }
+
+  "when primary stream fails, overall stream fails and inner stream is terminated" {
+    checkAll(Arb.throwable()) { e ->
+      val s = Stream.effect { Semaphore(0) }
+        .flatMap { semaphore ->
+          Stream(0)
+            .append { Stream.raiseError<Int>(e).delayBy(10.milliseconds) }
+            .switchMap {
+              Stream.effect { sleep(10.milliseconds) }.repeat().onFinalize { semaphore.release() }
+            }
+            .onFinalize { semaphore.acquire() }
+        }
+        .compile()
+
+      assertThrowable {
+        s.drain()
+      } shouldBe e
+    }
+  }
+
+  "when inner stream fails, inner stream finalizer run before the primary one" {
+    checkAll(Arb.stream(Arb.int()), Arb.throwable()) { s0, e ->
+      val s = Stream(0).append { s0 }
+
+      val s2 = Stream.effect {
+        Promise<Boolean>() to Atomic(false)
+      }.flatMap { (verdict, innerReleased) ->
+        s.delayBy(25.milliseconds)
```

Comment:
Do we need this `delayBy`? I'm not sure I understand how it affects the outcome of the tests.

Reply:
To be honest me neither, but at the beginning I was seeing more fails with the delays that without, so I presume it should do something indeed(?)